### DISCUSSION TITLE
Creation mfem::Mesh from gpu buffers

### DISCUSSION THREAD

**Discussion by moroz3339** [mfem/mfem#3465] [Q&A]
Hello everyone!
I construct a mesh as a list of triangles on a CUDA and next I must transfer it to Host for mfem::Mesh constructing, it seems wasteful because as the backend of MFEM, I use CUDA. So, is there any possibility of passing the CUDA pointer to GPU buffers of vertices, indices etc, for creating MFEM mesh?
Thanks!

**Comment by v-dobrev** [comment]
Hello,

Currently, almost all `mfem::Mesh` data lives on the CPU and almost all of its methods are executed on CPU even when using MFEM's `cuda` backend. One exception is the computation of the geometric factors in `Mesh::GetGeometricFactors` and `Mesh::GetFaceGeometricFactors` which is performed on GPU (when using a GPU backend). Basically, even if your mesh data is available directly on the GPU, we'll need to copy it to the CPU to construct the respective `mfem::Mesh` object.

In most cases, the main computational cost is not in the `Mesh` class, so we have not focused on porting more methods to the GPU, at least for now.
